### PR TITLE
Refactor KubeProxy to allow mocking of all moving parts.

### PR DESCRIPTION
**Pull request by gmarek**
Needed for #13731

@thockin - can you please take a quick look? I'm going to go through it one more time tomorrow, but I'd gladly take some comments. The main goal of this PR is to make it possible to run "HollowProxy" in a similar fashion to how "SimpleKubelet" is run.

cc @fgrzadkowski @wojtek-t @davidopp

### CODE REVIEW

File: cmd/hyperkube/kube-proxy.go [hunk 1]
```diff
@@ -19,24 +19,35 @@ limitations under the License.
...
 	hks := Server{
 		SimpleUsage: "proxy",
 		Long: `The Kubernetes proxy server is responsible for taking traffic directed at
-		services and forwarding it to the appropriate pods.  It generally runs on
-		nodes next to the Kubelet and proxies traffic from local pods to remote pods.
-		It is also used when handling incoming external traffic.`,
-		Run: func(_ *Server, args []string) error {
-			return s.Run(args)
-		},
+	 	services and forwarding it to the appropriate pods.  It generally runs on
```
Comment (karlkfi) [inline cmd/hyperkube/kube-proxy.go]: looks like this needs a go fmt.

**Comment by karlkfi** [comment]
Looks like a decent iterative enhancement to extract the config object, but I worry that NewProxyServer still has way too many branches. I'd like to see it cleaned up even more, maybe with multiple constructors and/or extracted helper methods.

I'm also a little worried that this doesn't come with any tests changes. Are there sufficient unit/integration tests to cover this refactor?

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -181,72 +177,141 @@ func (s *ProxyServer) Run(_ []string) error {
...
+	fs.StringVar(&s.HostnameOverride, "hostname-override", s.HostnameOverride, "If non-empty, will use this string as identification instead of the actual hostname.")
+	fs.BoolVar(&s.ForceUserspaceProxy, "legacy-userspace-proxy", true, "Use the legacy userspace proxy (instead of the pure iptables proxy).")
+	fs.DurationVar(&s.SyncPeriod, "iptables-sync-period", 5*time.Second, "How often iptables rules are refreshed (e.g. '5s', '1m', '2h22m').  Must be greater than 0.")
+	fs.BoolVar(&s.MasqueradeAll, "masquerade-all", false, "If using the pure iptables proxy, SNAT everything")
+	fs.BoolVar(&s.CleanupAndExit, "cleanup-iptables", false, "If true cleanup iptables rules and exit.")
+}
+
+// Run runs the specified ProxyServer.  This should never exit (unless CleanupAndExit is set).
+func (s *ProxyServer) Run(_ []string) error {
+
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: wasted line

**Comment by thockin** [comment]
I'd appreciate if you rebased onto https://github.com/kubernetes/kubernetes/pull/14310, which is all cleared to go it any minute now...

File: cmd/hyperkube/kube-proxy.go (hunk 1 above)
Comment (gmarek) [inline cmd/hyperkube/kube-proxy.go]: I wasn't touching this lines, and I don't really understand what I'm supposed to do. Deleting superfluous space for now.

**Comment by gmarek** [comment]
Yeah - I'd also like to see deeper refactor to allow easy mocking, but I'd rather have someone responsible for KubeProxy do it, as I'm just visiting:)

**Comment by gmarek** [comment]
@thockin @karlkfi PTAL

**Comment by karlkfi** [comment]
It's not clear to me that the NewProxyServerWithOverrides constructor abstraction is actually better...
 
Is there a way to process the config up front or use a proxy config to populate the required nil fields so that the constructor just uses them without branching?

It's also not clear to me why pointers to instantiated objects are in an object called "Config". I would expect a config to be stateless, but this one contains OomAdjuster, kube.Client, IptUtils (which is a bad name by itself). Having the naming not match the contents is a good sign that we need to not just extract config, but extract behavior.

What if the constructor takes all the overridable objects as fields and a wrapper constructor creates all the defaults? A normal user would use the wrapper, and the test would used the unwrapped constructor with more arguments.

**Comment by gmarek** [comment]
SG, I'll change it tomorrow.

**Comment by gmarek** [comment]
@karlkfi PTAL - I created a 'full' constructor, and changed 'old' one to the default one. It'll make mocking single things for tests harder, but it'll also keep the codebase cleaner, so I'm OK with this.

File: cmd/kube-proxy/app/server.go [hunk 3]
```diff
@@ -110,87 +109,110 @@ func checkKnownProxyMode(proxyMode string) bool {
...
+		EndpointsHandler: endpointsHandler,
+		IptInterface:     iptInterface,
+		OomAdjuster:      oomAdjuster,
+		Proxier:          proxier,
+		Recorder:         recorder,
+		ServiceConfig:    serviceConfig,
+	}, nil
+}
+
+// NewProxyServerWithOverrides creates a new ProxyServer object with default parameters but
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: comment and function name don't match

File: cmd/kube-proxy/app/server.go [hunk 4]
```diff
@@ -110,87 +109,110 @@ func checkKnownProxyMode(proxyMode string) bool {
...
+func NewProxyServer(
+	config *ProxyServerConfig,
+	client *kubeclient.Client,
+	endpointsConfig *proxyconfig.EndpointsConfig,
+	endpointsHandler proxyconfig.EndpointsConfigHandler,
+	iptInterface utiliptables.Interface,
+	oomAdjuster *oom.OomAdjuster,
+	proxier proxy.ProxyProvider,
+	recorder record.EventRecorder,
+	serviceConfig *proxyconfig.ServiceConfig) (*ProxyServer, error) {
```
Comment (karlkfi) [inline cmd/kube-proxy/app/server.go]: it's conventional to add a trailing comma and line break before the closing paren

File: cmd/kube-proxy/app/server.go [hunk 5]
```diff
@@ -110,87 +109,110 @@ func checkKnownProxyMode(proxyMode string) bool {
...
-		Namespace: "",
-	}
-
-	serviceConfig := config.NewServiceConfig()
-	endpointsConfig := config.NewEndpointsConfig()
-
-	var proxier proxy.ProxyProvider
-	var endpointsHandler config.EndpointsConfigHandler
-
+	// Create a IP table utils.
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: "iptables" not "IP tables"

File: cmd/kube-proxy/app/server.go (hunk 3 above)
Comment (gmarek) [inline cmd/kube-proxy/app/server.go]: Sorry, I forgot to update it after the change requested by Karl.

File: cmd/kube-proxy/app/server.go (hunk 5 above)
Comment (gmarek) [inline cmd/kube-proxy/app/server.go]: Done.

File: cmd/kube-proxy/app/server.go (hunk 4 above)
Comment (gmarek) [inline cmd/kube-proxy/app/server.go]: Done.

File: cmd/kube-proxy/app/server.go [hunk 6]
```diff
@@ -201,16 +223,16 @@ func (s *ProxyServer) Run(_ []string) error {
...
+		proxierIptables, err := iptables.NewProxier(iptInterface, execer, config.SyncPeriod, config.MasqueradeAll)
 		if err != nil {
 			glog.Fatalf("Unable to create proxier: %v", err)
 		}
 		proxier = proxierIptables
 		endpointsHandler = proxierIptables
 		// No turning back. Remove artifacts that might still exist from the userspace Proxier.
-		glog.V(2).Info("Tearing down userspace rules. Errors here are acceptable.")
-		userspace.CleanupLeftovers(ipt)
+		glog.V(2).Info("Tearing down userspace rules.Config. Errors here are acceptable.")
```
Comment (thockin) [inline cmd/kube-proxy/app/server.go]: I don't understand this line change - was it a search-replace artifact?

**Comment by thockin** [comment]
I've scanned it over a few times.  It's convoluted but it's a reasonable transformation.  I think we should be cleaner overall in the future - things like the "default" constructor are noisy.

**Comment by thockin** [comment]
TL;DR OK

File: cmd/kube-proxy/app/server.go (hunk 6 above)
Comment (gmarek) [inline cmd/kube-proxy/app/server.go]: Yes, sorry - I thought I fixed that...

**Comment by karlkfi** [comment]
I'm with @thockin on this one. Something about it makes me wish I had time to clean it up more, but it's a fine iterative improvement to add the flexibility you need.

I suspect once you try to use it you'll want to refactor it again. It's always hard adding abstraction layers without having multiple users.

LGTM

**Comment by karlkfi** [comment]
integration tests timed out on shippable. kicking it.

**Comment by gmarek** [comment]
I'm with both of you - it's pretty ugly and I just don't have enough resources to spend on cleaning it up properly (which would probably end pretty much equivalent to rewrite).
